master/internal/stream/supervisor.go

```diff
@@ -0,0 +1,150 @@
+	if err != nil {
+		log.Errorf("unable to get permission change listener: %s", err)
+		// XXX: investigate better recovery mechanism
+		return err
+	}
+	if permListener == nil {
+		// no listener means we don't have permissions configured at all
+		return nil
+	}
+	defer func() {
+		err := permListener.Close()
+		if err != nil {
+			log.Debugf("error occurred while closing permission listener: %s", err)
+		}
+	}()
+
+	for {
+		select {
+		// did permissions change?
+		case <-permListener.Notify:
+			log.Debugf("permission change detected, booting streamers")
+			func() {
+				ps.permissionLock.Lock()
+				defer ps.permissionLock.Unlock()
+				close(ps.permissionChangeChan)
+				ps.permissionChangeChan = make(chan struct{})
+			}()
+		// is the listener still alive?
+		case <-time.After(30 * time.Second):
+			pingErrChan := make(chan error)
+			go func() {
+				err = permListener.Ping()
+				pingErrChan <- errors.Wrap(err, "no active connection")
+			}()
+			if err := <-pingErrChan; err != nil {
+				log.Errorf("permission listener failed %s", err)
+				return err
+			}
+		// are we canceled?
+		case <-ctx.Done():
+			return nil
+		}
+	}
+}
+
+func (ssup *Supervisor) runOne(ctx context.Context) error {
+	group := errgroupx.WithContext(ctx)
+	subctx := group.Context()
+	func() {
+		ssup.cond.L.Lock()
+		defer ssup.cond.L.Unlock()
+		ssup.cond.Broadcast()
+
+		ssup.subctx = subctx
+		ssup.ps = NewPublisherSet()
+
+		// start monitoring permissions
+		group.Go(
+			func(c context.Context) error {
+				return monitorPermissionChanges(c, ssup.ps)
+			},
+		)
+		// start up all publishers
+		// XXX: should these be passing errors backup?
+		group.Go(ssup.ps.Start)
+		// group.Go(PublishLoop(ctx, "experiment_trial_chan", ps.Experiments))
+	}()
+	return group.Wait()
+}
+
+// Run attempts to start up the publisher system and recovers in the event of a failure.
+func (ssup *Supervisor) Run(ctx context.Context) error {
+	backoffSettings := backoff.NewExponentialBackOff()
+	backoffSettings.InitialInterval = minInterval
+	backoffSettings.MaxInterval = maxInterval
+
+	run := func() error {
+		err := ssup.runOne(ctx)
+		if err != nil {
+			log.Errorf("restarting publisher system after failure: %s", err)
+		}
+		return err
+	}
+
+	err := backoff.Retry(run, backoff.WithContext(backoffSettings, ctx))
+
+	if err != nil && err != backoff.Permanent(err) {
+		return errors.Wrap(err, "maximum number of retries reached")
+	}
+	return err
+}
+
+// Websocket is the method that we pass to the Echo server, because rb doesn't know how to update
+// the backing function for an Echo.GET() call.  So we need something that will live as long as
+// the master.
+func (ssup *Supervisor) Websocket(socket *websocket.Conn, c echo.Context) error {
+	var ssupCtx context.Context
+	var ps *PublisherSet
+	func() {
+		ssup.cond.L.Lock()
+		defer ssup.cond.L.Unlock()
+		for ssup.ps == nil {
+			ssup.cond.Wait()
+			break
+		}
+		ssupCtx = ssup.subctx
+		ps = ssup.ps
```

Comment:
I was thinking more about this, and I don't know if this is the right behavior.

I think it is great behavior if we assume the publisher will never fail.

But if the publisher ever gets into a state where it's totally toast, this will essentially cause all websockets to block forever (or until the client decides to give up on them).

It might be better to
- close the websocket and say "sorry publishing is broken; try again later".  I can imagine how pissed some customer will be if that's all they get from their webui pages, but at least it'll be obvious failures.
- allow the websocket to proceed with a nil or closed `ssupCtx`, let the websocket do its initial streaming operations, and then die when it tries to transition to live streaming.  That is a more graceful failure because as long as the postgres connection is still working, customers have a "limp" mode to rely on if we have a bug in the publishing code.

I also like option B for the simple fact that when we ship the new feature, if it breaks due to a bug on our side there will be a lot of angry people at us, in particular, which I would like to avoid.  In principle I like A or even what you currently have because it's simple but for self preservation I might prefer B.

Reply:
I also like the idea of having a limp mode, in the event that this is getting used on webui pages having some sort of fall-back behavior, that while not ideal won't completely disable every page that uses it (assuming the webui doesn't also have their own fall-back behavior).

While simplicity is nice, I'm willing to sacrifice simplicity for more reliable behavior in this case. This will also more closely the original behavior of 'closedChan'

Reply:
Isn't this already doing option B? Since we aren't concerned about what state the `ssupCtx` is in, all we really need is a valid PublisherSet so we can use it's websocket function. Are you suggesting that instead we have an alternative to `ps.Websocket()` that gets used in the event that `ssupCtx` & `ps` doesn't get set? 

Or are you suggesting that inside of `ps.Websocket()`, doing the `closed` logic last, so anything that is pending can get flushed out?

Reply:
Hm, I suppose there's just one special case, which is if a connection begins before the first Supervisor runOne starts.  That case is the only reason that there's a sync.Cond in use, and the blocking on sync.Cond _looks_ like the behavior is that we wouldn't go into limp mode.

So maybe there's a better way of handling this special case which is more expressive to somebody reading the code.

```go
func NewSupervisor() *Supervisor {
	// start with a valid publisher set and valid supervisor context, but let the
    // context start in a canceled state, so that connections before the first
    // call to runOne() can't hit nil pointer errors.
    ctx, cancel_func := context.WithCancel(context.Background())
    cancel_func()
    ps := NewPublisherSet()
	return &Supervisor{ps: ps, ctx: ctx}
}
```

Note that you'll still need a Mutex, but not a Cond.

Reply:
I do like the idea of making the intention behind the sync clearer, swapped out the cond for a mutex